### ISSUE TITLE
scylla-tools cassandra-stress drop in performance between 0.12 and 0.11

### ISSUE THREAD

**Issue by slivne** [scylladb/scylla#602]
from @glommer 

scylla-tools to scylla-0.11, each of my clients can push 167k ops/sec, and many cores are handling requests.
moving to scylla-0.12, I get 32k ops/sec with only one core handling requests.

from @penberg 

 Perhaps it's the Java driver upgrade that we merged pretty late in the cycle?

**Comment by slivne** [comment]
update: same issue was detected in cassandra

https://issues.apache.org/jira/browse/CASSANDRA-9558

working out what needs to be changed

**Comment by slivne** [comment]
so this is what they did

https://github.com/apache/cassandra/commit/26bd5124efbc08a657e703b3f9776d9fd6ba6c22

the main part
```
     public void connect(ProtocolOptions.Compression compression) throws Exception
     {
+        PoolingOptions poolingOpts = new PoolingOptions();
+        poolingOpts.setCoreConnectionsPerHost(HostDistance.LOCAL, 8);
         Cluster.Builder clusterBuilder = Cluster.builder()
                                                 .addContactPoint(host)
                                                 .withPort(port)
+                                                .withPoolingOptions(poolingOpts)
+                                                .withProtocolVersion(ProtocolVersion.V2)
                                                 .withoutMetrics(); // The driver uses metrics 3 with conflict with our version


```
moving back to v2 is not an option I'll try todo the other change of 8 connections and benchmark

**Comment by slivne** [comment]
Doing the change they suggest indeed brings the performance back up - yet it requires chaning the protocol to v3 as well.

The issue is a difference in the driver handling V2,V3

https://docs.datastax.com/en/drivers/java/2.1/com/datastax/driver/core/PoolingOptions.html

Extract

*With ProtocolVersion#V2 or below, there are at most 128 simultaneous requests per connection, so the pool defaults to a variable size. You will typically raise the maximum capacity by adding more connections with setMaxConnectionsPerHost(HostDistance, int).*

*With ProtocolVersion#V3 or above, there are up to 32768 requests per connection, and the pool defaults to a fixed size of 1. You will typically raise the maximum capacity by allowing more simultaneous requests per connection (setMaxRequestsPerConnection(HostDistance, int)). *

So the driver does not allow (in current version 2.1.8) to open multiple connections to the same end point as the number of parallel requests is higher - and the user is supposed to use that - so we are using a single connection and that is the reason for the performance drop - there is no way to work around this - if we are using v3 - and the original driver.

Three options:
1. Support V2 end to end - so that we can do what they are doing
2. Support V2 for non user profile cases (default) and with user profile use V3 - that will be much slower
3. Create our own version of driver that uses v3 and that will remove this limit on the number of connections

**Comment by slivne** [comment]
patching the driver to use the same limit of connections for v3 as in v2 raises the performance and applying the change for cassandra-stress that increases the number of connections solves the issue.

```
diff --git a/driver-core/src/main/java/com/datastax/driver/core/PoolingOptions.java b/driver-core/src/main/java/com/datastax/driver/core/PoolingOptions.java
index 4fd1b3b..e958516 100644
--- a/driver-core/src/main/java/com/datastax/driver/core/PoolingOptions.java
+++ b/driver-core/src/main/java/com/datastax/driver/core/PoolingOptions.java
@@ -65,10 +65,10 @@ public class PoolingOptions {
             .build(),
 
         ProtocolVersion.V3, ImmutableMap.<String, Integer>builder()
-            .put("corePoolLocal", 1)
-            .put("maxPoolLocal", 1)
+            .put("corePoolLocal", 2)
+            .put("maxPoolLocal", 8)
             .put("corePoolRemote", 1)
-            .put("maxPoolRemote", 1)
+            .put("maxPoolRemote", 2)
             .put("newConnectionThresholdLocal", 800)
             .put("newConnectionThresholdRemote", 200)
             .put("maxRequestsPerConnectionLocal", 1024)
diff --git a/driver-core/src/test/java/com/datastax/driver/core/PoolingOptionsTest.java b/driver-core/src/test/java/com/datastax/driver/core/PoolingOptionsTest.java
index c593679..846f729 100644
--- a/driver-core/src/test/java/com/datastax/driver/core/PoolingOptionsTest.java
+++ b/driver-core/src/test/java/com/datastax/driver/core/PoolingOptionsTest.java
@@ -45,10 +45,10 @@ public class PoolingOptionsTest {
         PoolingOptions options = new PoolingOptions();
         options.setProtocolVersion(ProtocolVersion.V3);
 
-        assertThat(options.getCoreConnectionsPerHost(LOCAL)).isEqualTo(1);
-        assertThat(options.getMaxConnectionsPerHost(LOCAL)).isEqualTo(1);
+        assertThat(options.getCoreConnectionsPerHost(LOCAL)).isEqualTo(2);
+        assertThat(options.getMaxConnectionsPerHost(LOCAL)).isEqualTo(8);
         assertThat(options.getCoreConnectionsPerHost(REMOTE)).isEqualTo(1);
-        assertThat(options.getMaxConnectionsPerHost(REMOTE)).isEqualTo(1);
+        assertThat(options.getMaxConnectionsPerHost(REMOTE)).isEqualTo(2);
         assertThat(options.getNewConnectionThreshold(LOCAL)).isEqualTo(800);
         assertThat(options.getNewConnectionThreshold(REMOTE)).isEqualTo(200);
         assertThat(options.getMaxRequestsPerConnection(LOCAL)).isEqualTo(1024);
@@ -65,7 +65,7 @@ public class PoolingOptionsTest {
         assertThat(options.getMaxConnectionsPerHost(LOCAL)).isEqualTo(8);
 
         // KO for v3 (default max = 1)
-        options = new PoolingOptions().setCoreConnectionsPerHost(LOCAL, 3);
+        options = new PoolingOptions().setCoreConnectionsPerHost(LOCAL, 9);
         try {
             options.setProtocolVersion(ProtocolVersion.V3);
             fail("Expected an IllegalArgumentException");
```

**Comment by slivne** [comment]
Some more background info

https://datastax-oss.atlassian.net/browse/JAVA-738

They did run a benchmark, according to their benchmark when the client is stressed limited (c3.large) there is a benefit for limiting the number of connections (at least in cassandra's case). When the client is not limited (c3.8xlarge) then multiple connections are better.

The more relevant lines are:

java738-v3 - cassandra-2.1.6 branch with 8 core connections and protocol v3 using driver branch java738.
java738-v3-1c - cassandra-2.1.6 branch with 1 core connections and protocol v3 using driver branch java738.


Bottom line with V3 if the client is stressed working with cassandra it seems better to limit the number of connections.
